### PR TITLE
Enable Oracle Lint and additional compiler warnings

### PR DESCRIPTION
**Pull request by johnsonjh**
* Enable Oracle Lint in CI
* Fix typo in `SUNPRO_C` macro
* Pass with additional warnings enabled …
  * GCC: 
    * `-Wformat-truncation=2`
    *  `-Wformat-overflow`
    *  `-Wformat=2`
    *  `-Wbad-function-cast` 
  * Clang: 
    * `-Wdisabled-macro-expansion` 
    * `-Wassign-enum` 
    * `-Wformat=2`
    *  `-Wshift-overflow` 
    * `-Wstring-conversion`

### CODE REVIEW

File: include/sir/config.h [hunk 1]
```diff
@@ -277,7 +277,7 @@
  * data in any color mode (the largest possible sequence, which is:
  * `\x1b[a;fb;m;rrr;ggg;bbb;fb;m;rrr;ggg;bbbm`) plus a null terminator.
  */
-# define SIR_MAXSTYLE 39
+# define SIR_MAXSTYLE 43
```
Comment (aremmell) [inline include/sir/config.h:280]: Where'd you come up with this?

File: src/sirhelpers.c [hunk 2]
```diff
@@ -30,7 +30,7 @@ bool __sir_validptrptr(const void* restrict* pp, bool fail) {
     bool valid = NULL != pp;
     if (!valid && fail) {
         (void)_sir_seterror(_SIR_E_NULLPTR);
-        SIR_ASSERT("!NULL pointer");
+        SIR_ASSERT(!(bool)"NULL pointer!");
```
Comment (aremmell) [inline src/sirhelpers.c:33]: I don't think you need the (bool) at least I have never used it unless one of your psycho checks doesn't let it through

File: src/sirhelpers.c [hunk 3]
```diff
@@ -30,7 +30,7 @@ bool __sir_validptrptr(const void* restrict* pp, bool fail) {
     bool valid = NULL != pp;
     if (!valid && fail) {
         (void)_sir_seterror(_SIR_E_NULLPTR);
-        SIR_ASSERT("!NULL pointer");
```
Comment (aremmell) [inline src/sirhelpers.c:33]: I do this all the time

File: src/sirhelpers.c [hunk 4]
```diff
@@ -39,7 +39,7 @@ bool __sir_validptr(const void* restrict p, bool fail) {
     bool valid = NULL != p;
     if (!valid && fail) {
         (void)_sir_seterror(_SIR_E_NULLPTR);
-        SIR_ASSERT(!"NULL pointer");
+        SIR_ASSERT(!(bool)"NULL pointer!");
```
Comment (aremmell) [inline src/sirhelpers.c:42]: I see you did this in multiple places. What's the warning?

File: src/sirinternal.c [hunk 5]
```diff
@@ -575,7 +575,7 @@ bool _sir_logv(sir_level level, PRINTF_FORMAT const char* format, va_list args)
         fmt = _sir_formattime(now, buf.timestamp, SIR_TIMEFORMAT);
         SIR_ASSERT_UNUSED(fmt, fmt);
 
-        (void)snprintf(buf.msec, SIR_MAXMSEC, SIR_MSECFORMAT, nowmsec);
```
Comment (aremmell) [inline src/sirinternal.c:578]: This warned with Wformat-truncation?

File: tests/tests.c [hunk 6]
```diff
@@ -720,7 +720,14 @@ bool sirtest_textstylesanity(void) {
     bool pass = si_init;
 
     printf("\t" WHITEB("--- explicitly invalid ---") "\n");
+#if defined(__clang__) || defined(__clang_version__)
+# pragma clang diagnostic push
+# pragma clang diagnostic ignored "-Wassign-enum"
+#endif
     pass &= !sir_settextstyle(SIRL_INFO, 0xbbb, 800, 920);
```
Comment (aremmell) [inline tests/tests.c:727]: What if you cast it to the enum type?

File: tests/tests.c [hunk 7]
```diff
@@ -1023,7 +1037,14 @@ bool sirtest_levelssanity(void) {
     printf(INDENT_ITEM WHITE("greater than SIRL_ALL: %04"PRIx16) "\n", invalid);
 
     /* individual invalid level. */
+#if defined(__clang__) || defined(__clang_version__)
+# pragma clang diagnostic push
+# pragma clang diagnostic ignored "-Wassign-enum"
+#endif
     sir_level invalid2 = 0x1337;
```
Comment (aremmell) [inline tests/tests.c:1044]: Same thing with these. Does a cast squash the warning? I really, really hate these pragma blocks.

Comment (aremmell) [reply tests/tests.c:1044]: Casting resolves that warning on AppleClang.

File: include/sir/config.h (hunk 1 above)
Comment (johnsonjh) [reply include/sir/config.h:280]: GCC did.  I can change it back and let CI fail so you can see the exact message.

File: tests/tests.c (hunk 7 above)
Comment (johnsonjh) [reply tests/tests.c:1043]: Yeah, that'll work

File: tests/tests.c (hunk 6 above)
Comment (johnsonjh) [reply tests/tests.c:727]: Will do

File: src/sirinternal.c (hunk 5 above)
Comment (johnsonjh) [reply src/sirinternal.c:578]: Yep.

File: src/sirhelpers.c (hunk 4 above)
Comment (johnsonjh) [reply src/sirhelpers.c:42]:
```sh
src/sirhelpers.c:42:21: error: implicit conversion turns string literal into bool: 'char[14]' to 'bool' [-Werror,-Wstring-conversion]
        SIR_ASSERT(!"NULL pointer!");
                   ~^~~~~~~~~~~~~~~
```

File: include/sir/config.h (hunk 1 above)
Comment (johnsonjh) [reply include/sir/config.h]: Here's the error: https://gitlab.com/libsir/libsir/-/jobs/4960927984#L70

File: plugins/sample/plugin_sample.c [hunk 8]
```diff
@@ -57,22 +57,22 @@ PLUGIN_EXPORT bool sir_plugin_query(sir_plugininfo* info) {
...
-    printf("\t" DGRAY("plugin_sample (%s): level: %04"PRIx32", message: %s"),
-        __func__, level, message);
+    (void)printf("\t" DGRAY("plugin_sample (%s): level: %04"PRIx32", message: %s"),
+                 __func__, level, message);
     return true;
 }
 
 PLUGIN_EXPORT bool sir_plugin_cleanup(void) { //-V524
-    printf("\t" DGRAY("plugin_sample ('%s')") "\n", __func__);
+    (void)printf("\t" DGRAY("plugin_sample ('%s')") "\n", __func__);
```
Comment (aremmell) [inline plugins/sample/plugin_sample.c:76]: which compiler warning flag caused this to be marked as a warning?

File: src/sirhelpers.c (hunk 2 above)
Comment (aremmell) [reply src/sirhelpers.c:33]: which compiler warning flag caused this to be marked as a warning?

**Review by aremmell — COMMENTED** [review]
Want to test these errors myself, so can you please just comment which compiler and warning flag I need to use for each one I commented on?

File: plugins/sample/plugin_sample.c [hunk 9]
```diff
@@ -23,10 +23,10 @@
  * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  */
+#include <stdio.h>
```
Comment (aremmell) [inline plugins/sample/plugin_sample.c:26]: Which compiler warning flag caused this one?

File: plugins/sample/plugin_sample.c (hunk 8 above)
Comment (johnsonjh) [reply plugins/sample/plugin_sample.c:76]: That's Oracle Lint

File: plugins/sample/plugin_sample.c (hunk 9 above)
Comment (johnsonjh) [reply plugins/sample/plugin_sample.c:26]: Not a warning per se on this one, but making the order the same as it is in the dummy plug-in.

**Comment by johnsonjh** [comment]
> Want to test these errors myself, so can you please just comment which compiler and warning flag I need to use for each one I commented on?

Just use `./.lint.sh extra && ./.lint.sh gccextra` from this branch with GCC 13.2.1 20230728 and Clang 16.0.6.  Both are in the base Docker image (`docker run --rm -it --name "temp-$(env TZ=UTC date -u '+%s')" registry.gitlab.com/libsir/libsir/base:latest` to run a temporary session)

File: src/sirhelpers.c (hunk 2 above)
Comment (aremmell) [reply src/sirhelpers.c:33]: n/m, it's string-conversion

File: src/sirinternal.c (hunk 5 above)
Comment (aremmell) [reply src/sirinternal.c:582]: Let me guess, it says 11 or 12 chars

File: src/sirfilesystem.c [hunk 10]
```diff
@@ -235,7 +235,7 @@ char* _sir_getappfilename(void) {
 
 #if !defined(__WIN__)
 # if defined(__READLINK_OS__)
-        /* Flawfinder: ignore */
+        /* Flawfinder: ignore */ /* LINTED E_SEC_READLINK_WARN */
```
[the remark this answers is not in the document]
Comment (johnsonjh) [reply src/sirfilesystem.c]: Those aren't there anymore

File: include/sir/platform.h [hunk 11]
```diff
@@ -523,4 +527,13 @@ typedef BOOL(CALLBACK* sir_once_fn)(PINIT_ONCE, PVOID, PVOID*);
 
 #include "sir/impl.h"
 
+/* Support Clang's -Wdisabled-macro-expansion check on Linux/glibc */
```
[the remark this answers is not in the document]
Comment (johnsonjh) [reply include/sir/platform.h:530]: IMG_8180

Undefining the recursive macro is what [other](https://github.com/forthy42/gforth/blob/651151683fd5e5d404d89b81b821b7983fa6999b/unix/i2c.i#L8) [software](https://github.com/jevolk/charybdis/blob/52ed73c81d9e2ea958a90abb7ac996fd500d70c2/include/ircd/stdinc.h#L127C1-L127C1) [does](https://github.com/wxPerl/wxPerl/blob/b094a88cd5d9d8f4bb159eb7176905428c3d4eeb/cpp/wxapi.h#L138), and is required to run this test.  I'll wrap it in a lint-specific definition though, so it won't affect normal builds (though the compiler output is actually the same regardless of the definition.)

anyway, conditionally compiled out in commit 315661e

File: src/sirhelpers.c [hunk 12]
```diff
@@ -244,7 +244,7 @@ bool __sir_validstr(const char* restrict str, bool fail) {
     bool valid = str && (*str != '\0');
     if (!valid && fail) {
         (void)_sir_seterror(_SIR_E_STRING);
-        SIR_ASSERT(!"invalid string");
+        SIR_ASSERT(!(bool)"invalid string!");
```
[the remark this answers is not in the document]
Comment (johnsonjh) [reply src/sirhelpers.c]: Done in commit 2706ec6

File: src/sirhelpers.c [hunk 13]
```diff
@@ -121,7 +121,7 @@ bool _sir_validupdatedata(sir_update_config_data* data) {
 
     if (!valid) {
         (void)_sir_seterror(_SIR_E_INVALID);
-        SIR_ASSERT("!invalid sir_update_config_data");
```
[the remark this answers is not in the document]
Comment (johnsonjh) [reply src/sirhelpers.c:121]: Done in commit 2706ec6

**Review by aremmell — APPROVED** [review]

**Pull request merged**